stdlib/asm/crypto/dsa/falcon.masm

```diff
@@ -1,278 +1,611 @@
+    drop drop
+    #=> [...]
 end
 
-#! Given four elements on stack top, this routine computes squared norm of that
-#! vector ( read polynomial ) with four coefficients.
-#!
-#! Imagine, given vector is f, which is described as
-#!
-#! f = [a0, a1, a2, a3]
-#!
-#! Norm of that vector is
+
+# SQUARE NORM OF Z_q[x]/(phi) POLYNOMIALS 
+# =============================================================================================
+
+#! Normalizes an `e` in [0, q) to be in [-(q-1) << 1, (q-1) << 1) and returns its square norm.
 #!
-#! √(a0 ^ 2 + a1 ^ 2 + a2 ^ 2 + a3 ^ 2)
+#! We use the following formula to do so:
+#! normalize(e) = e^2 - phi * (2*q*e - q^2) where phi := (e > (q - 1)/2)
 #!
-#! But we need squared norm, which is just skipping the final square root operation.
+#! The formula implements:
 #!
-#! Expected stack state :
+#! if e > (q-1)/2:
+#!   return (q - e)^2
+#! else:
+#!   return e^2 
 #!
-#! [a0, a1, a2, a3, ...]
+#! The use of the formula avoids using the if-else block.
 #!
-#! Final stack state :
+#! Input: [e, ...]
+#! Output [norm(e)^2, ...]
 #!
-#! [b, ...] | b = a0 ^ 2 + a1 ^ 2 + a2 ^ 2 + a3 ^ 2
-proc.squared_norm_word
+#! Cycles: 21
+export.norm_sq
+    
     dup
-    mul
-
+    dup mul
     swap
-    dup
-    mul
+    #=> [e, e^2, ...]
 
-    add
-
-    swap
     dup
-    mul
-
-    add
+    push.6144
+    u32unchecked_gt
+    #=> [phi, e, e^2]
 
     swap
-    dup
-    mul
+    mul.24578       # 2*q 
+    push.151019521  # q^2
+    sub
+    #=> [2*q*e - q^2, phi, e^2
 
-    add
+    mul sub
+    #=> [norm(e)^2, ...]
 end
 
-#! Given a degree 512 polynomial in coefficient form, as starting (absolute) memory address
-#! on stack, this routine computes squared norm of that vector, using following formula
-#!
-#! Say, f = [a0, a1, a2, ..., a510, a511]
-#!      g = sq_norm(f) = a0 ^ 2 + a1 ^ 2 + ... + a510 ^ 2 + a511 ^ 2
+#! On input a tuple (u, w, v), the following computes (v - (u + (- w % q) % q) % q).
+#! We can avoid doing three modular reductions by using the following facts:
+#! 
+#! 1. q is much smaller than the Miden prime. Precisely, q * 2^50 < Q 
+#! 2. The coefficients of the product polynomial, u and w, are less than J := 512 * q^2
+#! 3. The coefficients of c are less than q.
 #!
-#! Expected input stack state :
+#! This means that we can substitute (v - (u + (- w % q) % q) % q) with  v + w + J - u without
+#! risking Q-overflow since |v + w + J - u| < 1025 * q^2
 #!
-#! [f_start_addr, ...] | f_addr`i` holds f[(i << 2) .. ((i+1) << 2)]
+#! To get the final result we reduce (v + w + J - u) modulo q.
 #!
-#! Consecutive 127 addresses on stack can be computed using `INCR` instruction, because memory
-#! addresses are consecutive i.e. monotonically increasing by 1.
+#! Input: [v, w, u, ...]
+#! Output: [e, ...]
 #!
-#! Final stack state :
-#!
-#! [g, ...] | g = sq_norm(f)
-export.squared_norm_poly512
-    push.0.0.0.0.0
+#! Cycles: 44
+export.diff_mod_q
+
+    # 1) v + w + J 
+    add add.77321994752
```

Comment:
nit: would be good to replace this value with a constant.

Reply:
Done!